### PR TITLE
refactor: fail build on compiler warnings

### PR DESCRIPTION
**Pull request by fractalwrench**
## Goal

Detect compiler warnings introduced in changes to Java code quickly.

## Changeset

- Updated gradle to error on any warnings found during compilation
- Address existing issues (raw types, missing `serialVersionUID`
- Suppress unchecked cast/deprecation warnings

## Tests

Ran tests on CI

## Review

For the submitter, initial self-review:

- [ ] Commented on code changes inline explain the reasoning behind the approach
- [ ] Reviewed the test cases added for completeness and possible points for discussion
- [ ] A changelog entry was added for the goal of this pull request
- [ ] Check the scope of the changeset - is everything in the diff required for the pull request?
- This pull request is ready for:
  - [ ] Initial review of the intended approach, not yet feature complete
  - [ ] Structural review of the classes, functions, and properties modified
  - [ ] Final review

For the pull request reviewer(s), this changeset has been reviewed for:

- [ ] Consistency across platforms for structures or concepts added or modified
- [ ] Consistency between the changeset and the goal stated above
- [ ] Internal consistency with the rest of the library - is there any overlap between existing interfaces and any which have been added?
- [ ] Usage friction - is the proposed change in usage cumbersome or complicated?
- [ ] Performance and complexity - are there any cases of unexpected O(n^3) when iterating, recursing, flat mapping, etc?
- [ ] Concurrency concerns - if components are accessed asynchronously, what issues will arise
- [ ] Thoroughness of added tests and any missing edge cases
- [ ] Idiomatic use of the language

### CODE REVIEW

File: src/test/java/com/bugsnag/BugsnagTest.java [hunk 1]
```diff
@@ -339,6 +342,7 @@ public void beforeNotify(Report report) {
         assertFalse(bugsnag.notify(new Throwable()));
     }
 
+    @SuppressWarnings("deprecation")
```
Comment (simonbowring) [inline src/test/java/com/bugsnag/BugsnagTest.java]: Probably worth a comment that it's the deprecated method that's under test. We should also add a test `testEndpoints` that tests `setEndpoints`.

File: build.gradle [hunk 2]
```diff
@@ -44,3 +44,10 @@ test {
 checkstyle {
     toolVersion = "6.16"
 }
+
+gradle.projectsEvaluated {
+    tasks.withType(JavaCompile) {
+        // fail build on warnings, disable options complaining about Java 6 compatibility when building with JDK 7+
```
Comment (simonbowring) [inline build.gradle]: If we are supporting JDK 6 then I'm not sure that suppressing the Java 6 compatibility warning is something that we should do. If we want to guarantee compatibility then we should set `options.bootClasspath` to point to a JDK6 `rt.jar`.

On the other hand, I realize that this means that a JDK 6 installation is required to build which might not be practical. If you think it's fine to suppress this then we should look at getting JDK 6 CI sorted.

File: src/test/java/com/bugsnag/BugsnagTest.java (hunk 1 above)
Comment (fractalwrench) [reply src/test/java/com/bugsnag/BugsnagTest.java]: `ConfigurationTest` covers `setEndpoints`, I'll update with an inline comment

File: build.gradle (hunk 2 above)
Comment (fractalwrench) [reply build.gradle]: Fair point - I'll see if we can prioritise getting JDK 6 building on CI, or think about whether we need to support it still.
